base: Allow pausing and resuming facmod and buildmod (12.10.20)

dd84ai opened the pull request:
pause and resume ability for module and device construction
new commands
/base facmod switch <Index>//pauses if active, resumes if paused
/base buildmod switch <index> //pauses if active, resumes if paused

dsyalex commented on Plugins/Public/base_plugin/PlayerCommands.cpp:
```diff
@@ -810,6 +810,37 @@ namespace PlayerCommands
 			base->Save();
 			PrintUserCmdText(client, L"OK Module destroyed");
 		}
+		else if (cmd == L"switch")
```
switch is a weird command name for this

dd84ai replied:
remade as "pause" and "resume" separate commands

dsyalex commented on Plugins/Public/base_plugin/FactoryModule.cpp:
```diff
@@ -273,16 +273,16 @@ bool FactoryModule::ClearQueue()
 }
 
-bool FactoryModule::PauseResumeQueue()
+bool FactoryModule::PauseQueue(bool NewState)
```
how about ToggleQueuePaused

dd84ai replied:
sounds good

dsyalex commented on Plugins/Public/base_plugin/FactoryModule.cpp:
```diff
@@ -273,16 +273,16 @@ bool FactoryModule::ClearQueue()
 {
-	if (Paused)
+	bool RememberState = Paused;
+	if (NewState)
```
Paused = NewState;

dd84ai replied:
ok

dsyalex commented on Plugins/Public/base_plugin/PlayerCommands.cpp:
```diff
@@ -965,6 +1081,7 @@ namespace PlayerCommands
 			PrintUserCmdText(client, L"|  list - show factory modules and build status");
 			PrintUserCmdText(client, L"|  clear <index> - clear build queue for factory module at <index>");
 			PrintUserCmdText(client, L"|  add <index> <type> - add item <type> to build queue for factory module at <index>");
+			PrintUserCmdText(client, L"|  switch <index> - pause/resume factory module at <index>");
```
needs updating

dd84ai replied:
i did it

dsyalex commented on Plugins/Public/base_plugin/PlayerCommands.cpp:
```diff
@@ -965,6 +1081,8 @@ namespace PlayerCommands
 			PrintUserCmdText(client, L"|  list - show factory modules and build status");
 			PrintUserCmdText(client, L"|  clear <index> - clear build queue for factory module at <index>");
 			PrintUserCmdText(client, L"|  add <index> <type> - add item <type> to build queue for factory module at <index>");
+			PrintUserCmdText(client, L"|  pause <index> - pause factory module at <index>");
```
should we move these down to below the type definitions that seem to be associated with the `add` doc above?

dd84ai replied:
mmm....
pause and resume look to be the best under "add" because they can be used only after "add" was applied
at the same time all commands look to be better placed at the top of instruction, keeping all commands easily foundable grouped at its beginning

btw I noticed they write | at phrase start to fix issue with phrase disappearance. I'll add the same to "autobuy" to have indents

dsyalex replied:
they can be under add, but right now they appear between add's main docs and the docs for one of its parameters. please put pause/resume after the docs for <type>

dd84ai replied:
if you insist

dsyalex reviewed (commented):
little thing inline, otherwise good to go

dsyalex commented:
oh, also can we squash please

dd84ai commented:
addressed

dsyalex reviewed (commented):
Can we have a better commit message please

dsyalex reviewed (changes requested):
(setting review state)

dsyalex reviewed (dismissed):
(needs rebase again, sorry)

dd84ai commented:
Performing tests:
1) build construct two modules, paused, they aren't consuming
2) Reloaded module, pause perssisted.
2) added resources, unpaused. Modules started consuming resources
3) Added third module docking factory, paused first two modules. Only third module is constructing 
4) attempt to pause not buildmodule, failed
5) Attempt to craft device, success
6) Attempt to pause not factory, failed
7) Attempt to resume factory, success

All tests have been passed
